src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java

```diff
@@ -0,0 +1,148 @@
+
+        // initialize step listeners to chain steps
+        StepListener<Void> createIndexStep = new StepListener<>();
+        StepListener<String> createPipelineStep = new StepListener<>();
+        StepListener<Void> indexFeatureStep = new StepListener<>();
+        StepListener<Exception> deletePipelineStep = new StepListener<>();
+
+        if (isIndexAlreadyExists) {
+            createIndexStep.onResponse(null); // mark create index step as completed, to continue to next steps.
+        } else {
+            // create index
+            Map<String, String> fieldMap = new HashMap<>();
+            fieldMap.put(content.getFieldName(), content.getFieldType());
+            indexManager.create(content.getIndexName(), Collections.unmodifiableMap(fieldMap), createIndexStep);
+        }
+        // create a pipeline after creating index
+        createIndexStep.whenComplete(
+            notUsed -> pipelineManager.create(content.getFieldName(), createPipelineStep),
+            flowListener::onFailure
+        );
+
+        // index features as document after creating pipeline
+        createPipelineStep.whenComplete(
+            pipeline -> { indexContentAsDocument(pipeline, content, indexFeatureStep); },
+            createPipelineFailedException -> { flowListener.onFailure(createPipelineFailedException); }
+        );
+
+        // delete pipeline
+        indexFeatureStep.whenComplete(notUsed -> {
+            String pipeline = createPipelineStep.result();
+            pipelineManager.delete(pipeline, deletePipelineStep, () -> null);
+        }, uploadFailed -> {
+            String pipeline = createPipelineStep.result();
+            pipelineManager.delete(pipeline, deletePipelineStep, () -> uploadFailed);
+        });
+
+        // set response or failure depending on previous steps status
+        deletePipelineStep.whenComplete(uploadFailedException -> {
+            if (uploadFailedException != null) {
+                throw uploadFailedException;
+            }
+            flowListener.onResponse(new AcknowledgedResponse(true));
+        }, flowListener::onFailure);
+
+    }
+
+    private void indexContentAsDocument(String pipeline, UploadGeoJSONRequestContent content, StepListener<Void> uploadStepListener) {
+        Optional<BulkRequestBuilder> contentRequestBuilder = contentBuilder.prepare(content, pipeline);
+        if (!contentRequestBuilder.isPresent()) {
+            uploadStepListener.onFailure(new IllegalStateException("No valid features are available to index"));
+            return;
+        }
+        contentRequestBuilder.get().execute(ActionListener.wrap(bulkResponse -> {
+            if (bulkResponse.hasFailures()) {
+                final String failureMessage = buildUploadFailureMessage(bulkResponse);
+                throw new IllegalStateException(failureMessage);
+            }
+            LOGGER.info("indexed " + bulkResponse.getItems().length + " features");
```

Comment:
Sorry forgot to respond. I added Index name in Index creation step. We are not setting index name to BulkRequest, only way to get index name is to iterate the response. I feel it will add more latency to iterate just for log message.